crates/matrix-sdk/src/client/builder.rs

```diff
@@ -641,3 +738,273 @@ impl ClientBuildError {
+    // rather than the plain server name, otherwise the service will prepend https:// to the name
+    // and the request will fail. In practice, this isn't a problem as the service
+    // first strips the scheme and then checks if the name is a valid server
+    // name, so it is a close enough approximation.
+
+    #[async_test]
+    async fn test_discovery_invalid_server() {
+        // Given a new client builder.
+        let mut builder = ClientBuilder::new();
+
+        // When building a client with an invalid server name.
+        builder = builder.server_name_or_homeserver_url("⚠️ This won't work 🚫");
+        let error = builder.build().await.unwrap_err();
+
+        // Then the operation should fail due to the invalid server name.
+        assert!(matches!(error, ClientBuildError::InvalidServerName));
+    }
+
+    #[async_test]
+    async fn test_discovery_no_server() {
+        // Given a new client builder.
+        let mut builder = ClientBuilder::new();
+
+        // When building a client with a valid server name that doesn't exist.
+        builder = builder.server_name_or_homeserver_url("localhost:3456");
+        let error = builder.build().await.unwrap_err();
+
+        // Then the operation should fail with an HTTP error.
+        println!("{error}");
+        assert!(matches!(error, ClientBuildError::Http(_)));
+    }
+
+    #[async_test]
+    async fn test_discovery_web_server() {
+        // Given a random web server that isn't a Matrix homeserver or hosting the
+        // well-known file for one.
+        let server = MockServer::start().await;
+        let mut builder = ClientBuilder::new();
+
+        // When building a client with the server's URL.
+        builder = builder.server_name_or_homeserver_url(server.uri());
+        let error = builder.build().await.unwrap_err();
+
+        // Then the operation should fail with a server discovery error.
+        assert!(matches!(error, ClientBuildError::AutoDiscovery(FromHttpResponseError::Server(_))));
+    }
+
+    #[async_test]
+    async fn test_discovery_direct_legacy() {
+        // Given a homeserver without a well-known file.
+        let homeserver = make_mock_homeserver().await;
+        let mut builder = ClientBuilder::new();
+
+        // When building a client with the server's URL.
+        builder = builder.server_name_or_homeserver_url(homeserver.uri());
+        let client = builder.build().await.unwrap();
+
+        // Then a client should be built without support for sliding sync or OIDC.
+        #[cfg(feature = "experimental-sliding-sync")]
+        assert_eq!(client.sliding_sync_proxy(), None);
+        #[cfg(feature = "experimental-oidc")]
+        assert!(client.oidc().authentication_server_info().is_none());
+    }
+
+    #[async_test]
+    async fn test_discovery_direct_legacy_custom_proxy() {
+        // Given a homeserver without a well-known file and a service with a custom
```

Comment:
```suggestion
        // Given a homeserver without a well-known file and with a custom
```

Reply:
Can you tell where these tests started? 😅